internal/resource/equality_test.go

```diff
@@ -283,6 +283,310 @@ func TestEqualID(t *testing.T) {
+	})
+
+	t.Run("nil", func(t *testing.T) {
+		a := &pbresource.Reference{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name:    "qux",
+			Section: "blah",
+		}
+		b := &pbresource.ID{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name: "qux",
+			Uid:  ulid.Make().String(),
+		}
+		require.False(t, resource.ReferenceOrIDMatch(a, nil))
+		require.False(t, resource.ReferenceOrIDMatch(nil, b))
+	})
+
+	t.Run("different type", func(t *testing.T) {
+		a := &pbresource.Reference{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name:    "qux",
+			Section: "blah",
+		}
+		b := &pbresource.ID{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name: "qux",
+			Uid:  ulid.Make().String(),
+		}
+		b.Type.Kind = "album"
+		require.False(t, resource.ReferenceOrIDMatch(a, b))
+	})
+
+	t.Run("different tenancy", func(t *testing.T) {
+		a := &pbresource.Reference{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name:    "qux",
+			Section: "blah",
+		}
+		b := &pbresource.ID{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name: "qux",
+			Uid:  ulid.Make().String(),
+		}
+		b.Tenancy.Namespace = "qux"
+		require.False(t, resource.ReferenceOrIDMatch(a, b))
+	})
+
+	t.Run("different name", func(t *testing.T) {
+		a := &pbresource.Reference{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name:    "qux",
+			Section: "blah",
+		}
+		b := &pbresource.ID{
+			Type: &pbresource.Type{
+				Group:        "demo",
+				GroupVersion: "v2",
+				Kind:         "artist",
+			},
+			Tenancy: &pbresource.Tenancy{
+				Partition: "foo",
+				PeerName:  "bar",
+				Namespace: "baz",
+			},
+			Name: "qux",
+			Uid:  ulid.Make().String(),
+		}
+		b.Name = "boom"
+		require.False(t, resource.ReferenceOrIDMatch(a, b))
+	})
```

Comment:
Ah actually I had wanted this function to do reference comparison of only the type/tenancy/name fields. It only accepts the common interface between the two, so  it would not have visibility into the section (ref) or uid (id) fields.

Perhaps there's an alternate function name we could use to suggest that?  The interface is currently called `ReferenceOrID` hence the current function name.

Reply:
Ah ok. I think a comment is fine if there isn't one already.